### PR TITLE
#5635: Serial/OpenMP:  Parallel_scan with return value for TeamThreadRange

### PR DESCRIPTION
**Pull request by thearusable**
Related to #5635

### CODE REVIEW

**Comment by fnrizzi** [comment]
please improve description

**Comment by fnrizzi** [comment]
i think we are missing some guards in the test, aren't we? how about openmp and threads? 
https://kokkos.github.io/kokkos-core-wiki/keywords.html#device-backends

File: core/src/impl/Kokkos_HostThreadTeam.hpp [hunk 1]
```diff
@@ -864,55 +864,79 @@ KOKKOS_INLINE_FUNCTION
 
 //----------------------------------------------------------------------------
 
-template <typename iType, class Closure, class Member>
+template <typename iType, class Closure, class Member,
+          class ValueType = typename Kokkos::Impl::FunctorAnalysis<
+              Kokkos::Impl::FunctorPatternInterface::SCAN, void, Closure,
+              void>::value_type>
```
Comment (masterleinad) [inline core/src/impl/Kokkos_HostThreadTeam.hpp:870]: Why does `ValueType` need a default type? It should be deduced from the third function argument, right?
It makes more sense to have a `static_assert` checking the compatibility of the return value's type and the functor's value type.
```suggestion
          class ValueType>
```

File: core/src/impl/Kokkos_HostThreadTeam.hpp [hunk 2]
```diff
@@ -864,55 +864,79 @@ KOKKOS_INLINE_FUNCTION
...
 
-  value_type scan_val = value_type();
+  ValueType scan_val = ValueType();
+  parallel_scan(loop_boundaries, closure, scan_val);
+}
 
+template <typename iType, class ClosureType, class Member,
+          class ValueType = typename Kokkos::Impl::FunctorAnalysis<
+              Impl::FunctorPatternInterface::SCAN, void, ClosureType,
+              void>::value_type>
```
Comment (masterleinad) [inline core/src/impl/Kokkos_HostThreadTeam.hpp:910]: Same here
```suggestion
          class ValueType>
```

File: core/unit_test/incremental/Test16_ParallelScan.hpp [hunk 3]
```diff
@@ -104,6 +104,164 @@ struct TestScan {
       ASSERT_EQ(h_data(i), upd);
     }
   }
+
+// Temporary: This condition will progressively be reduced when parallel_scan
+// with return value will be implemented for more backends.
+#if defined(KOKKOS_ENABLE_SERIAL) || defined(KOKKOS_ENABLE_OPENMP)
+#if !defined(KOKKOS_ENABLE_CUDA) && !defined(KOKKOS_ENABLE_HIP) &&             \
+    !defined(KOKKOS_ENABLE_OPENACC) && !defined(KOKKOS_ENABLE_SYCL) &&         \
+    !defined(KOKKOS_ENABLE_THREADS) && !defined(KOKKOS_ENABLE_OPENMPTARGET) && \
```
Comment (masterleinad) [inline core/unit_test/incremental/Test16_ParallelScan.hpp:113]: This should also work for `Threads`, right?
```suggestion
    !defined(KOKKOS_ENABLE_OPENMPTARGET) && \
```

**Review by masterleinad — CHANGES_REQUESTED** [review]

File: core/unit_test/incremental/Test16_ParallelScan.hpp (hunk 3 above)
Comment (thearusable) [reply core/unit_test/incremental/Test16_ParallelScan.hpp:113]: No, I checked and for `Threads` there are additional changes required. There will be a separate PR for `Threads`

File: core/src/impl/Kokkos_HostThreadTeam.hpp (hunk 1 above)
Comment (thearusable) [reply core/src/impl/Kokkos_HostThreadTeam.hpp]: That's good point. I updated that to use `static_assert` to check compatibility of the types. In current changes I needed to add a default value for `FunctorAnalysis` because it couldn't deduce the type when there was a generic functor used. (`GenericExclusiveScanFunctor` from the tests)

File: core/src/impl/Kokkos_HostThreadTeam.hpp (hunk 2 above)
Comment (thearusable) [reply core/src/impl/Kokkos_HostThreadTeam.hpp]: Same as above

**Review by masterleinad — APPROVED** [review]
Fine with me.

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/impl/Kokkos_HostThreadTeam.hpp [hunk 4]
```diff
@@ -864,55 +864,87 @@ KOKKOS_INLINE_FUNCTION
...
-  value_type accum = 0;
+      ValueType>::value_type;
+  static_assert(std::is_same<ClosureValueType, ValueType>::value,
+                "Non-matching value types of closure and return type");
 
   // Intra-member scan
   for (iType i = loop_boundaries.start; i < loop_boundaries.end;
        i += loop_boundaries.increment) {
-    closure(i, accum, false);
+    closure(i, return_val, false);
```
Comment (dalg24) [inline core/src/impl/Kokkos_HostThreadTeam.hpp:884]: That does not look right.  The Kokkos semantics for reductions and scan with return value is we ignore the "initial" value of the argument that is passed by the user and overwrite with the result.

File: core/unit_test/incremental/Test16_ParallelScan.hpp
Comment (dalg24) [inline core/unit_test/incremental/Test16_ParallelScan.hpp:1]: Why did you decide to add to the incremental tests instead of the regular unit tests.  These are intended to guide the development of new backends and the feature you are contributing is an advanced feature that I don't necessarily expect to see in the incremental tests.

File: core/src/impl/Kokkos_HostThreadTeam.hpp (hunk 4 above)
Comment (fnrizzi) [reply core/src/impl/Kokkos_HostThreadTeam.hpp:884]: @thearusable here is the link: https://kokkos.github.io/kokkos-core-wiki/API/core/parallel-dispatch/parallel_scan.html#semantics

i think we need to use a local variable `accum` as before, and then assign 
```
value_type accum = 0;
... 
return_value = accum; 
```

File: core/src/impl/Kokkos_HostThreadTeam.hpp [hunk 5]
```diff
@@ -864,55 +864,87 @@ KOKKOS_INLINE_FUNCTION
...
 
-  // 'accum' output is the exclusive prefix sum
-  accum = loop_boundaries.thread.team_scan(accum);
+  // 'return_val' output is the exclusive prefix sum
+  return_val = loop_boundaries.thread.team_scan(return_val);
 
   for (iType i = loop_boundaries.start; i < loop_boundaries.end;
        i += loop_boundaries.increment) {
-    closure(i, accum, true);
+    closure(i, return_val, true);
```
Comment (fnrizzi) [inline core/src/impl/Kokkos_HostThreadTeam.hpp:892]: I think we need to use an `accum` local variable as before, then at the end assign to `return_value`

File: core/src/impl/Kokkos_HostThreadTeam.hpp [hunk 6]
```diff
@@ -864,55 +864,87 @@ KOKKOS_INLINE_FUNCTION
...
-  value_type scan_val = value_type();
+  ValueType scan_val = ValueType();
+  parallel_scan(loop_boundaries, closure, scan_val);
+}
+
+template <typename iType, class ClosureType, class Member, typename ValueType>
+KOKKOS_INLINE_FUNCTION
+    std::enable_if_t<!Kokkos::is_reducer<ValueType>::value &&
+                     Impl::is_host_thread_team_member<Member>::value>
+    parallel_scan(Impl::ThreadVectorRangeBoundariesStruct<iType, Member> const&
```
Comment (fnrizzi) [inline core/src/impl/Kokkos_HostThreadTeam.hpp:915]: I think this overload should go into a separate PR

File: core/src/impl/Kokkos_HostThreadTeam.hpp [hunk 7]
```diff
@@ -864,55 +864,87 @@ KOKKOS_INLINE_FUNCTION
...
        i += loop_boundaries.increment) {
-    closure(i, scan_val, true);
+    closure(i, return_val, true);
   }
 }
 
+template <typename iType, class ClosureType, class Member>
+KOKKOS_INLINE_FUNCTION
+    std::enable_if_t<Impl::is_host_thread_team_member<Member>::value>
+    parallel_scan(Impl::ThreadVectorRangeBoundariesStruct<iType, Member> const&
```
Comment (fnrizzi) [inline core/src/impl/Kokkos_HostThreadTeam.hpp:937]: i think this should go into a separate PR

**Review by fnrizzi — CHANGES_REQUESTED** [review]

File: core/src/impl/Kokkos_HostThreadTeam.hpp [hunk 8]
```diff
@@ -891,6 +893,23 @@ KOKKOS_INLINE_FUNCTION
...
+    std::enable_if_t<Impl::is_host_thread_team_member<Member>::value>
+    parallel_scan(Impl::TeamThreadRangeBoundariesStruct<iType, Member> const&
+                      loop_boundaries,
+                  Closure const& closure) {
+  // Extract ValueType from the closure
+  using ValueType = typename Kokkos::Impl::FunctorAnalysis<
+      Kokkos::Impl::FunctorPatternInterface::SCAN, void, Closure,
+      void>::value_type;
+
+  ValueType scan_val = ValueType();
```
Comment (masterleinad) [inline core/src/impl/Kokkos_HostThreadTeam.hpp:911]:
```suggestion
  ValueType scan_val;
```

**Review by masterleinad — CHANGES_REQUESTED** [review]

File: core/src/impl/Kokkos_HostThreadTeam.hpp (hunk 6 above)
Comment (thearusable) [reply core/src/impl/Kokkos_HostThreadTeam.hpp]: Ok I created a separate PR for that: https://github.com/kokkos/kokkos/pull/6292

File: core/src/impl/Kokkos_HostThreadTeam.hpp (hunk 7 above)
Comment (thearusable) [reply core/src/impl/Kokkos_HostThreadTeam.hpp]: Ok I created a separate PR for that: https://github.com/kokkos/kokkos/pull/6292

File: core/src/impl/Kokkos_HostThreadTeam.hpp (hunk 4 above)
Comment (thearusable) [reply core/src/impl/Kokkos_HostThreadTeam.hpp]: Ok, Fixed

File: core/unit_test/incremental/Test16_ParallelScan.hpp
Comment (thearusable) [reply core/unit_test/incremental/Test16_ParallelScan.hpp:1]: Thanks for the clarification! At first I thought that incremental testing might be a good place to add my tests, but it actually doesn't make much sense now.

Now they are splitted and added to `TestTeamScan` and `TestTeamVector`.

File: core/src/impl/Kokkos_HostThreadTeam.hpp (hunk 5 above)
Comment (thearusable) [reply core/src/impl/Kokkos_HostThreadTeam.hpp]: Ok, Fixed

File: core/src/impl/Kokkos_HostThreadTeam.hpp (hunk 8 above)
Comment (thearusable) [reply core/src/impl/Kokkos_HostThreadTeam.hpp]: Ok, fixed

**Review by masterleinad — APPROVED** [review]

**Review by fnrizzi — APPROVED** [review]

**Comment by fnrizzi** [comment]
retest this please

**Comment by fnrizzi** [comment]
@thearusable please make sure you also post a draft PR in the wiki to update the documentation on the parallel_scan interface to reflect this PR

File: core/unit_test/incremental/Test16_ParallelScan.hpp [hunk 9]
```diff
@@ -152,4 +152,4 @@ TEST(TEST_CATEGORY, IncrTest_16_parallelscan) {
   test_total.parallel_scan<GenericExclusiveScanFunctor<TEST_EXECSPACE>>();
 }
 
-}  // namespace Test
+}  // namespace Test
```
Comment (fnrizzi) [inline core/unit_test/incremental/Test16_ParallelScan.hpp:155]: this needs to be fixed

**Review by fnrizzi — CHANGES_REQUESTED** [review]
see comment

File: core/unit_test/incremental/Test16_ParallelScan.hpp (hunk 9 above)
Comment (thearusable) [reply core/unit_test/incremental/Test16_ParallelScan.hpp]: Fixed

**Review by fnrizzi — APPROVED** [review]

**Comment by fnrizzi** [comment]
retest this please

**Comment by fnrizzi** [comment]
@thearusable this still have failures

for example:
```
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:240: Failure
4: The difference between scan_ref and a_os(i) is 495, which exceeds abs_err, where
4: scan_ref evaluates to 561,
4: a_os(i) evaluates to 66, and
4: abs_err evaluates to 4.0531158447265625e-06.
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:240: Failure
4: The difference between scan_ref and a_os(i) is 1521, where
4: scan_ref evaluates to 1711,
4: a_os(i) evaluates to 190.
4: The abs_error parameter abs_err evaluates to 1.3100631690576847e-14 which is smaller than the minimum distance between doubles for numbers of this magnitude which is 2.8421709430404007e-14, thus making this EXPECT_NEAR check equivalent to EXPECT_EQUAL. Consider using EXPECT_DOUBLE_EQ instead.
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:240: Failure
4: The difference between scan_ref and a_os(i) is 1849, which exceeds abs_err, where
4: scan_ref evaluates to 2080,
4: a_os(i) evaluates to 231, and
4: abs_err evaluates to 7.7486038208007812e-06.
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:240: Failure
4: The difference between scan_ref and a_os(i) is 61009, where
4: scan_ref evaluates to 68635,
4: a_os(i) evaluates to 7626.
4: The abs_error parameter abs_err evaluates to 8.2378548427186615e-14 which is smaller than the minimum distance between doubles for numbers of this magnitude which is 9.0949470177292824e-13, thus making this EXPECT_NEAR check equivalent to EXPECT_EQUAL. Consider using EXPECT_DOUBLE_EQ instead.
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:238: Failure
4: Expected equality of these values:
4:   scan_ref
4:     Which is: 486591
4:   a_os(i)
4:     Which is: 53956
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:240: Failure
4: The difference between scan_ref and a_os(i) is 763439, where
4: scan_ref evaluates to 858705,
4: a_os(i) evaluates to 95266.
4: The abs_error parameter abs_err evaluates to 2.9110047705671604e-13 which is smaller than the minimum distance between doubles for numbers of this magnitude which is 1.4551915228366852e-11, thus making this EXPECT_NEAR check equivalent to EXPECT_EQUAL. Consider using EXPECT_DOUBLE_EQ instead.
4: [  FAILED  ] openmp.team_scan_ret_val (1001 ms)
```

**Comment by masterleinad** [comment]
Please rebase on top of develop to fix some CI issues.

**Comment by thearusable** [comment]
> @thearusable this still have failures
> 
> for example:
> 
> ```
> 4: /var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:240: Failure
> 4: The difference between scan_ref and a_os(i) is 495, which exceeds abs_err, where
> 4: scan_ref evaluates to 561,
> 4: a_os(i) evaluates to 66, and
> 4: abs_err evaluates to 4.0531158447265625e-06.
> 4: /var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:240: Failure
> 4: The difference between scan_ref and a_os(i) is 1521, where
> 4: scan_ref evaluates to 1711,
> 4: a_os(i) evaluates to 190.
> 4: The abs_error parameter abs_err evaluates to 1.3100631690576847e-14 which is smaller than the minimum distance between doubles for numbers of this magnitude which is 2.8421709430404007e-14, thus making this EXPECT_NEAR check equivalent to EXPECT_EQUAL. Consider using EXPECT_DOUBLE_EQ instead.
> 4: /var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:240: Failure
> 4: The difference between scan_ref and a_os(i) is 1849, which exceeds abs_err, where
> 4: scan_ref evaluates to 2080,
> 4: a_os(i) evaluates to 231, and
> 4: abs_err evaluates to 7.7486038208007812e-06.
> 4: /var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:240: Failure
> 4: The difference between scan_ref and a_os(i) is 61009, where
> 4: scan_ref evaluates to 68635,
> 4: a_os(i) evaluates to 7626.
> 4: The abs_error parameter abs_err evaluates to 8.2378548427186615e-14 which is smaller than the minimum distance between doubles for numbers of this magnitude which is 9.0949470177292824e-13, thus making this EXPECT_NEAR check equivalent to EXPECT_EQUAL. Consider using EXPECT_DOUBLE_EQ instead.
> 4: /var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:238: Failure
> 4: Expected equality of these values:
> 4:   scan_ref
> 4:     Which is: 486591
> 4:   a_os(i)
> 4:     Which is: 53956
> 4: /var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:240: Failure
> 4: The difference between scan_ref and a_os(i) is 763439, where
> 4: scan_ref evaluates to 858705,
> 4: a_os(i) evaluates to 95266.
> 4: The abs_error parameter abs_err evaluates to 2.9110047705671604e-13 which is smaller than the minimum distance between doubles for numbers of this magnitude which is 1.4551915228366852e-11, thus making this EXPECT_NEAR check equivalent to EXPECT_EQUAL. Consider using EXPECT_DOUBLE_EQ instead.
> 4: [  FAILED  ] openmp.team_scan_ret_val (1001 ms)
> ```

Ok, fixed

**Comment by thearusable** [comment]
> Please rebase on top of develop to fix some CI issues.

Ok, rebased

**Review by crtrott — APPROVED** [review]

**Comment by crtrott** [comment]
I am a bit concerned about NVHPC not having run, but will risk merging it anyway.
